tempesta_fw/cache.c

```diff
@@ -1847,54 +1837,124 @@ tfw_cache_purge_method(TfwHttpReq *req)
+ * set to avoid any data copies.
+ * - for https connections - pages are reused in skbs and SKBTX_SHARED_FRAG is
+ * set, but in-place crypto operations are not allowed, so data copy happens
+ * right before data is pushed into network.
+ * - for h2 connections - copying happens on constructing responses from
+ * cache, and SKBTX_SHARED_FRAG is left unset allowing in-place crypto
+ * operations.
+ *
+ * Since we can't encrypt shared data in-place we always copy it, so we need
+ * reserve some space in cached pages to avoid extra skb fragmentation. Since
+ * body fragments are stored in cache by pages with at least cache record
+ * header preceding, which is bigger than h2 frame header, it's always possible
+ * to fit body fragment and h2 header into a single page. So there is no need
+ * to actually reserve any space for h2 frame header.
+ */
+static int
+tfw_cache_add_body_page(TfwMsgIter *it, char *p, int sz, TfwFrameHdr *frame_hdr,
+			bool h2, bool last_frag)
+{
+	int off;
+	struct page *page;
+
+	/*
+	 * @sz is guarantied to be bigger than FRAME_HEADER_SIZE if frame header
+	 * is stored in cache before data. True for first fragment.
+	 */
+	if (h2) {
+		char *new_p;
+		if (!(page = alloc_page(GFP_ATOMIC))) {
+			return -ENOMEM;
+		}
+		new_p = page_address(page);
+		off = 0;
+		frame_hdr->flags = last_frag ? HTTP2_F_END_STREAM : 0;
+		frame_hdr->length = sz;
+		memcpy_fast(new_p + FRAME_HEADER_SIZE, p, sz);
+		sz += FRAME_HEADER_SIZE;
+		tfw_h2_pack_frame_header(new_p, frame_hdr);
+	}
+	else {
+		off = ((unsigned long)p & ~PAGE_MASK);
+		page = virt_to_page(p);
+	}
+
+	skb_fill_page_desc(it->skb, it->frag, page, off, sz);
+	skb_frag_ref(it->skb, it->frag);
+	ss_skb_adjust_data_len(it->skb, sz);
+	++it->frag;
+
+	return 0;
+}
+
 /**
  * Build the message body as paged fragments of skb.
  * See do_tcp_sendpages() as reference.
  */
 static int
-tfw_cache_build_resp_body(TDB *db, TdbVRec *trec, TfwMsgIter *it, char *p)
+tfw_cache_build_resp_body(TDB *db, TdbVRec *trec, TfwMsgIter *it, char *p,
+			  unsigned long body_sz, bool h2, unsigned int stream_id)
 {
 	int r;
+	TfwFrameHdr frame_hdr = {.stream_id = stream_id, .type = HTTP2_DATA};
 
 	if (WARN_ON_ONCE(!it->skb_head))
 		return -EINVAL;
 	/*
 	 * If all skbs/frags are used up (see @tfw_http_msg_expand_data()),
 	 * create new skb with empty frags to reference the cached body;
-	 * otherwise, use next empty frag in current skb.
+	 * otherwise, use next empty frag in current skb. Constructing h2
+	 * responses from cache always imply data copies, for SKBTX_SHARED_FRAG
+	 * skbs, copy will happen twice. If body is relatively small keep
+	 * using existing skb and allow double copy, otherwise allocate a new
+	 * skb even if previous is not fully populated.
 	 */
-	if (!it->skb) {
+	if (!it->skb || (++it->frag >= MAX_SKB_FRAGS)
+	    || (h2 && (skb_shinfo(it->skb)->tx_flags & SKBTX_SHARED_FRAG)
+		&& body_sz > PAGE_SIZE))
```

Comment:
> copy will happen twice

Looks like an unreasonable behaviour - why to do the 2 copies?

The other question is why do we ever see `SKBTX_SHARED_FRAG` flag in skbs which we build on our own? Typically we see the flag for proxied skbs or zero-copied data from the cache.

The 3rd question to address is that it seems at the moment the only place where we set the flag is `tfw_http_msg_expand_data()` and the function is called not only for HTTP/2, but for generic logic like `__tfw_http_add_hdr_via()`. Need to clarify with @aleksostapenko .

The last question to check is that in
```
$ grep SKBTX_SHARED_FRAG *.[ch]
cache.c: * - for http connections - pages are reused in skbs and SKBTX_SHARED_FRAG is
cache.c: * - for https connections - pages are reused in skbs and SKBTX_SHARED_FRAG is
cache.c: * cache, and SKBTX_SHARED_FRAG is left unset allowing in-place crypto
cache.c:	 * responses from cache always imply data copies, for SKBTX_SHARED_FRAG
cache.c:	    || (h2 && (skb_shinfo(it->skb)->tx_flags & SKBTX_SHARED_FRAG)
cache.c:			skb_shinfo(it->skb)->tx_flags &= ~SKBTX_SHARED_FRAG;
http_msg.c:			skb_shinfo(it->skb)->tx_flags |= SKBTX_SHARED_FRAG;
ss_skb.c:	if (skb_shinfo(skb)->tx_flags & SKBTX_SHARED_FRAG) {
ss_skb.c:		skb_shinfo(skb)->tx_flags &= ~SKBTX_SHARED_FRAG;
```
I don't see that the flag is set for zero-copied data from the cache for HTTP/1 transmission.

Reply:
Yeah, the function `tfw_http_msg_expand_data()` initially was intended only for transformations during H1->H2 forwarding. But later, during cache implementation I mistakenly began to use it also for response generation from cache including H1 responses. It seems that SKBTX_SHARED_FRAG flag in `tfw_http_msg_expand_data()` function should be set conditionally - e.g. depending on flag argument passed into it.

Reply:
@ikoveshnikov could you please fix this point as well? I think some refactoring is required: `tfw_http_msg_expand_data()` doesn't look like a proper place for setting `SKBTX_SHARED_FRAG`.

Reply:
Yes, this is exact thing I'm currently working on.

Reply:
I've reviewed, when we _need_ to set this flag, and it seems to be only case when a cached body is added into http/1 response.  `tfw_http_msg_expand_data()`  allocates a new skb and fragments and copies data there. ` SKBTX_SHARED_FRAG` is absolutely unreasonable for it.

 I reworked that and now `tfw_http_msg_expand_data()` doesn't manage `tx_flags`, instead the proper set of flags is set during cached body construction. `SKBTX_SHARED_FRAG`  is never set for cached response headers since we always copy them.